File: src/Application/Keyspace/Client/Python/keyspace.py
from keyspace_client import *

KEYSPACE_SUCCESS = 0
KEYSPACE_API_ERROR = -1

KEYSPACE_PARTIAL = -101
KEYSPACE_FAILURE = -102

KEYSPACE_NOMASTER = -201
KEYSPACE_NOCONNECTION =	-202

KEYSPACE_MASTER_TIMEOUT = -301
KEYSPACE_GLOBAL_TIMEOUT = -302

KEYSPACE_NOSERVICE = -401
KEYSPACE_FAILED = -402

def str_status(status):
	if status == KEYSPACE_SUCCESS:
		return "KEYSPACE_SUCCESS"
	elif status == KEYSPACE_API_ERROR:
		return "KEYSPACE_API_ERROR"
	elif status == KEYSPACE_PARTIAL:
		return "KEYSPACE_PARTIAL"
	elif status == KEYSPACE_FAILURE:
		return "KEYSPACE_FAILURE"
	elif status == KEYSPACE_NOMASTER:
		return "KEYSPACE_NOMASTER"
	elif status == KEYSPACE_NOCONNECTION:
		return "KEYSPACE_NOCONNECTION"
	elif status == KEYSPACE_MASTER_TIMEOUT:
		return "KEYSPACE_MASTER_TIMEOUT"
	elif status == KEYSPACE_GLOBAL_TIMEOUT:
		return "KEYSPACE_GLOBAL_TIMEOUT"
	elif status == KEYSPACE_NOSERVICE:
		return "KEYSPACE_NOSERVICE"
	elif status == KEYSPACE_FAILED:
		return "KEYSPACE_FAILED"

class Client:
	class Result:
		def __init__(self, cptr):
			self.cptr = cptr

		def __del__(self):
			self.close()
		
		def close(self):
			Keyspace_ResultClose(self.cptr)
			self.cptr = None
		
		def key(self):
			return Keyspace_ResultKey(self.cptr)
		
		def value(self):
			return Keyspace_ResultValue(self.cptr)
		
		def begin(self):
			return Keyspace_ResultBegin(self.cptr)
		
		def is_end(self):
			return Keyspace_ResultIsEnd(self.cptr)
		
		def next(self):
			return Keyspace_ResultNext(self.cptr)
		
		def command_status(self):
			return Keyspace_ResultCommandStatus(self.cptr)
		
		def transport_status(self):
			return Keyspace_ResultTransportStatus(self.cptr)
		
		def connectivity_status(self):
			return Keyspace_ResultConnectivityStatus(self.cptr)
		
		def timeout_status(self):
			return Keyspace_ResultTimeoutStatus(self.cptr)
		
		def key_values(self):
			kv = {}
			self.begin()
			while not self.is_end():
				kv[self.key()] = self.value()
				self.next()
			return kv
	
	def __init__(self, nodes):
		self.cptr = Keyspace_Create()
		self.result = None
		node_params = Keyspace_NodeParams(len(nodes))
		for node in nodes:
			node_params.AddNode(node)
		Keyspace_Init(self.cptr, node_params)
		node_params.Close()

	def __del__(self):
		del self.result
		Keyspace_Destroy(self.cptr)


	def set_global_timeout(self, timeout):
		Keyspace_SetGlobalTimeout(self.cptr, long(timeout))
	
	def get_global_timeout(self):
		return long(Keyspace_GetGlobalTimeout(self.cptr))
	
	def set_master_timeout(self, timeout):
		Keyspace_SetMasterTimeout(self.cptr, long(timeout))
	
	def get_master_timeout(self):
		return long(Keyspace_GetMasterTimeout(self.cptr))
	
	def get_master(self):
		return Keyspace_GetMaster(self.cptr)
	
	def get_state(self, node):
		return Keyspace_GetState(self.cptr, int(node))
	
	def distribute_dirty(self):
		return Keyspace_DistributeDirty(self.cptr)

	def get(self, key):
		status = Keyspace_Get(self.cptr, key)
		if status < 0:
			self.result = Client.Result(Keyspace_GetResult(self.cptr))
			return
		if Keyspace_IsBatched(self.cptr):
			return
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		return self.result.value()

	def dirty_get(self, key):
		status = Keyspace_DirtyGet(self.cptr, key)
		if status < 0:
			self.result = Client.Result(Keyspace_GetResult(self.cptr))
			return
		if Keyspace_IsBatched(self.cptr):
			return
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		return self.result.value()

	def count(self, prefix = "", start_key = "", count = 0, skip = False, forward = True):
		status = Keyspace_Count(self.cptr, prefix, start_key, count, skip, forward)
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		if status < 0:
			return None
		return int(self.result.value())

	def dirty_count(self, prefix = "", start_key = "", count = 0, skip = False, forward = True):
		status = Keyspace_DirtyCount(self.cptr, prefix, start_key, count, skip, forward)
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		if status < 0:
			return None
		return int(self.result.value())
	
	def list_keys(self, prefix = "", start_key = "", count = 0, skip = False, forward = True):
		status = Keyspace_ListKeys(self.cptr, prefix, start_key, count, skip, forward)
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		if status < 0:
			return None
		keys = []
		self.result.begin()
		while not self.result.is_end():
			keys.append(self.result.key())
			self.result.next()
		return keys
	
	def dirty_list_keys(self, prefix = "", start_key = "", count = 0, skip = False, forward = True):
		status = Keyspace_DirtyListKeys(self.cptr, prefix, start_key, count, skip, forward)
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		if status < 0:
			return None
		keys = []
		self.result.begin()
		while not self.result.is_end():
			keys.append(self.result.key())
			self.result.next()
		return keys
	
	def list_key_values(self, prefix = "", start_key = "", count = 0, skip = False, forward = True):
		status = Keyspace_ListKeyValues(self.cptr, prefix, start_key, count, skip, forward)
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		if status < 0:
			return None
		keyvals = {}
		self.result.begin()
		while not self.result.is_end():
			keyvals[self.result.key()] = self.result.value()
			self.result.next()
		return keyvals

	def dirty_list_key_values(self, prefix = "", start_key = "", count = 0, skip = False, forward = True):
		status = Keyspace_DirtyListKeyValues(self.cptr, prefix, start_key, count, skip, forward)
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		if status < 0:
			return None
		keyvals = {}
		self.result.begin()
		while not self.result.is_end():
			keyvals[self.result.key()] = self.result.value()
			self.result.next()
		return keyvals
	
	def set(self, key, value):
		status = Keyspace_Set(self.cptr, key, value)
		if status < 0:
			self.result = Client.Result(Keyspace_GetResult(self.cptr))
			return
		if Keyspace_IsBatched(self.cptr):
			return
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		return status

	def test_and_set(self, key, test, value):
		status = Keyspace_TestAndSet(self.cptr, key, test, value)
		if status < 0:
			self.result = Client.Result(Keyspace_GetResult(self.cptr))
			return
		if Keyspace_IsBatched(self.cptr):
			return
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		return self.result.value()
	
	def add(self, key, num):
		status = Keyspace_Add(self.cptr, key, long(num))
		if status < 0:
			self.result = Client.Result(Keyspace_GetResult(self.cptr))
			return
		if Keyspace_IsBatched(self.cptr):
			return
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		return int(self.result.value())
	
	def delete(self, key):
		status = Keyspace_Delete(self.cptr, key)
		if status < 0:
			self.result = Client.Result(Keyspace_GetResult(self.cptr))
			return
		if Keyspace_IsBatched(self.cptr):
			return
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		return status
	
	def remove(self, key):
		status = Keyspace_Remove(self.cptr, key)
		if status < 0:
			self.result = Client.Result(Keyspace_GetResult(self.cptr))
			return
		if Keyspace_IsBatched(self.cptr):
			return
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		return self.result.value()
	
	def rename(self, src, dst):
		status = Keyspace_Rename(self.cptr, src, dst)
		if status < 0:
			self.result = Client.Result(Keyspace_GetResult(self.cptr))
			return
		if Keyspace_IsBatched(self.cptr):
			return
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		return status
	
	def prune(self, prefix):
		status = Keyspace_Prune(self.cptr, prefix)
		if status < 0:
			self.result = Client.Result(Keyspace_GetResult(self.cptr))
			return
		if Keyspace_IsBatched(self.cptr):
			return
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		return status

	def set_expiry(self, key, exptime):
		status = Keyspace_SetExpiry(self.cptr, key, exptime)
		if status < 0:
			self.result = Client.Result(Keyspace_GetResult(self.cptr))
			return
		if Keyspace_IsBatched(self.cptr):
			return
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		return status
	
	def remove_expiry(self, key):
		status = Keyspace_RemoveExpiry(self.cptr, key)
		if status < 0:
			self.result = Client.Result(Keyspace_GetResult(self.cptr))
			return
		if Keyspace_IsBatched(self.cptr):
			return
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		return status
	
	def begin(self):
		return Keyspace_Begin(self.cptr)
	
	def submit(self):
		status = Keyspace_Submit(self.cptr)
		self.result = Client.Result(Keyspace_GetResult(self.cptr))
		return status
	
	def cancel(self):
		return Keyspace_Cancel(self.cptr)
		
	def _set_trace(self, trace):
		Keyspace_SetTrace(trace)

		

File: src/Application/Keyspace/Client/PHP/keyspace.php
<?php

include_once("keyspace_client.php");

define("KEYSPACE_SUCCESS", 0);
define("KEYSPACE_API_ERROR", -1);

define("KEYSPACE_PARTIAL", -101);
define("KEYSPACE_FAILURE", -102);

define("KEYSPACE_NOMASTER", -201);
define("KEYSPACE_NOCONNECTION", -202);

define("KEYSPACE_MASTER_TIMEOUT", -301);
define("KEYSPACE_GLOBAL_TIMEOUT", -302);

define("KEYSPACE_NOSERVICE", -401);
define("KEYSPACE_FAILED", -402);


class Result {
	private $cPtr;
	
	public function __construct($cptr) {
		$this->cPtr = $cptr;
	}
	
	public function __destruct() {
		$this->close();
	}
	
	public function close() {
		keyspace_client::Keyspace_ResultClose($this->cPtr);
		$this->cPtr = NULL;
	}
	
	public function key() {
		return keyspace_client::Keyspace_ResultKey($this->cPtr);
	}
	
	public function value() {
		return keyspace_client::Keyspace_ResultValue($this->cPtr);
	}
	
	public function begin() {
		return keyspace_client::Keyspace_ResultBegin($this->cPtr);
	}
	
	public function isEnd() {
		return keyspace_client::Keyspace_ResultIsEnd($this->cPtr);
	}
	
	public function next() {
		return keyspace_client::Keyspace_ResultNext($this->cPtr);
	}

	public function commandStatus() {
		return keyspace_client::Keyspace_ResultCommandStatus($this->cPtr);
	}

	public function transportStatus() {
		return keyspace_client::Keyspace_ResultTransportStatus($this->cPtr);
	}
	
	public function connectivityStatus() {
		return keyspace_client::Keyspace_ResultConnectivityStatus($this->cPtr);
	}
	
	public function timeoutStatus() {
		return keyspace_client::Keyspace_ResultTimeoutStatus($this->cPtr);
	}

	public function keys() {
		$keys = array();
		$this->begin();
		while (! $this->isEnd()) {
			$keys[] = $this->key();
			$this->next();
		}

		return $keys;
	}
	
	public function keyValues() {
		$kv = array();
		$this->begin();
		while (! $this->isEnd()) {
			$kv[$this->key()] = $this->value();
			$this->next();
		}

		return $kv;
	}
}

class KeyspaceClient {	
	private $co = NULL;
	public $result = NULL;
	
	public function __construct($nodes) {
		$this->co = keyspace_client::Keyspace_Create();
		$nodeParams = new Keyspace_NodeParams(count($nodes));
		foreach ($nodes as $node) {
			$nodeParams->AddNode($node);
		}
		keyspace_client::Keyspace_Init($this->co, $nodeParams);
		$nodeParams->Close();
	}
	
	public function __destruct() {
		keyspace_client::Keyspace_Destroy($this->co);
	}
	
	public static function statusToString($status) {
		switch ($status) {
		case KEYSPACE_SUCCESS:
	                return "KEYSPACE_SUCCESS";
        	case KEYSPACE_API_ERROR:
	                return "KEYSPACE_API_ERROR";
        	case KEYSPACE_PARTIAL:
	                return "KEYSPACE_PARTIAL";
        	case KEYSPACE_FAILURE:
	                return "KEYSPACE_FAILURE";
        	case KEYSPACE_NOMASTER:
	                return "KEYSPACE_NOMASTER";
		case KEYSPACE_NOCONNECTION:
	                return "KEYSPACE_NOCONNECTION";
        	case KEYSPACE_MASTER_TIMEOUT:
                	return "KEYSPACE_MASTER_TIMEOUT";
        	case KEYSPACE_GLOBAL_TIMEOUT:
	                return "KEYSPACE_GLOBAL_TIMEOUT";
        	case KEYSPACE_NOSERVICE:
                	return "KEYSPACE_NOSERVICE";
        	case KEYSPACE_FAILED:
	                return "KEYSPACE_FAILED";
		}

		return "<UNKNOWN>";
	}
	
	public function setGlobalTimeout($timeout) {
		keyspace_client::Keyspace_SetGlobalTimeout($this->co, $timeout);
	}
	
	public function getGlobalTimeout() {
		return keyspace_client::Keyspace_GetGlobalTimeout($this->co);
	}
	
	public function setMasterTimeout($timeout) {
		keyspace_client::Keyspace_SetMasterTimeout($this->co, $timeout);
	}
	
	public function getMasterTimeout() {
		return keyspace_client::Keyspace_GetMasterTimeout($this->co);
	}
	
	public function getMaster() {
		return keyspace_client::Keyspace_GetMaster($this->co);
	}
	
	public function getState() {
		return keyspace_client::Keyspace_GetState($this->co);
	}
	
	public function distributeDirty($dist) {
		return keyspace_client::Keyspace_DistributeDirty($this->co, $dist);
	}
	
	public function get($key) {
		$status = keyspace_client::Keyspace_Get($this->co, $key);
		if ($status < 0) {
			$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
			return NULL;
		}
		if ($this->isBatched())
			return NULL;
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
		return $this->result->value();
	}

	public function dirtyGet($key) {
		$status = keyspace_client::Keyspace_DirtyGet($this->co, $key);
		if ($status < 0) {
			$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
			return NULL;
		}
		if ($this->isBatched())
			return NULL;
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
		return $this->result->value();
	}

	public function count_($prefix = "", $start_key = "", $count = 0, $skip = FALSE, $forward = TRUE) {
		$status = keyspace_client::Keyspace_CountStr($this->co, $prefix, $start_key, $count, $skip, $forward);
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
		if ($status < 0)
			return NULL;
		return $this->result->value();
	}
	
	public function count($params) {
		$prefix = $this->listArg($params, "prefix");
		$start_key = $this->listArg($params, "start_key");
		$count = $this->listArg($params, "count");
		$skip = $this->listArg($params, "skip");
		$forward = $this->listArg($params, "forward");
		return $this->count_($prefix, $start_key, $count, $skip, $forward);
	}
	
	public function dirtyCount_($prefix = "", $start_key = "", $count = 0, $skip = FALSE, $forward = TRUE) {
		$status = keyspace_client::Keyspace_DirtyCountStr($this->co, $prefix, $start_key, $count, $skip, $forward);
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
		if ($status < 0)
			return NULL;
		return $this->result->value();
	}
	
	public function dirtyCount($params) {
		$prefix = $this->listArg($params, "prefix");
		$start_key = $this->listArg($params, "start_key");
		$count = $this->listArg($params, "count");
		$skip = $this->listArg($params, "skip");
		$forward = $this->listArg($params, "forward");
		return $this->dirtyCount_($prefix, $start_key, $count, $skip, $forward);
	}

	public function listKeys_($prefix = "", $start_key = "", $count = 0, $skip = FALSE, $forward = TRUE) {
		$status = keyspace_client::Keyspace_ListKeysStr($this->co, $prefix, $start_key, $count, $skip, $forward);
                $this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
                if ($status < 0)
                	return NULL;
		return $this->result->keys();
	}

	public function listKeys($params) {
		$prefix = $this->listArg($params, "prefix");
		$start_key = $this->listArg($params, "start_key");
		$count = $this->listArg($params, "count");
		$skip = $this->listArg($params, "skip");
		$forward = $this->listArg($params, "forward");
		return $this->listKeys_($prefix, $start_key, $count, $skip, $forward);
	}

	public function dirtyListKeys_($prefix = "", $start_key = "", $count = 0, $skip = FALSE, $forward = TRUE) {
		$status = keyspace_client::Keyspace_DirtyListKeysStr($this->co, $prefix, $start_key, $count, $skip, $forward);
                $this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
                if ($status < 0)
                	return NULL;
		return $this->result->keys();
	}
	
	public function dirtyListKeys($params) {
		$prefix = $this->listArg($params, "prefix");
		$start_key = $this->listArg($params, "start_key");
		$count = $this->listArg($params, "count");
		$skip = $this->listArg($params, "skip");
		$forward = $this->listArg($params, "forward");
		return $this->dirtyListKeys_($prefix, $start_key, $count, $skip, $forward);
	}

	public function listKeyValues_($prefix = "", $start_key = "", $count = 0, $skip = FALSE, $forward = TRUE) {
		$status = keyspace_client::Keyspace_ListKeyValuesStr($this->co, $prefix, $start_key, $count, $skip, $forward);
                $this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
                if ($status < 0)
                	return NULL;
		return $this->result->keyValues();
	}
	
	public function listKeyValues($params) {
		$prefix = $this->listArg($params, "prefix");
		$start_key = $this->listArg($params, "start_key");
		$count = $this->listArg($params, "count");
		$skip = $this->listArg($params, "skip");
		$forward = $this->listArg($params, "forward");
		return $this->listKeyValues_($prefix, $start_key, $count, $skip, $forward);
	}
	
	public function dirtyListKeyValues_($prefix = "", $start_key = "", $count = 0, $skip = FALSE, $forward = TRUE) {
		$status = keyspace_client::Keyspace_DirtyListKeyValues($this->co, $prefix, $start_key, $count, $skip, $forward);
                $this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
                if ($status < 0)
                	return NULL;
		return $this->result->keyValues();
	}
	
	public function dirtyListKeyValues($params) {
		$prefix = $this->listArg($params, "prefix");
		$start_key = $this->listArg($params, "start_key");
		$count = $this->listArg($params, "count");
		$skip = $this->listArg($params, "skip");
		$forward = $this->listArg($params, "forward");
		return $this->dirtyListKeyValues_($prefix, $start_key, $count, $skip, $forward);
	}
	
	public function set($key, $value) {
		$status = keyspace_client::Keyspace_Set($this->co, $key, $value);
		if ($status < 0) {
			$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
			return NULL;
		}
		if ($this->isBatched())
			return NULL;
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
	}

	public function testAndSet($key, $test, $value) {
		$status = keyspace_client::Keyspace_TestAndSet($this->co, $key, $test, $value);
		if ($status < 0) {
			$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
			return NULL;
		}
		if ($this->isBatched())
			return NULL;
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
		return $this->result->value();
	}

	public function add($key, $num) {
		$status = keyspace_client::Keyspace_AddStr($this->co, $key, $num);
		if ($status < 0) {
			$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
			return NULL;
		}
		if ($this->isBatched())
			return NULL;
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
		return $this->result->value();
	}
	
	public function delete($key) {
		$status = keyspace_client::Keyspace_Delete($this->co, $key);
		if ($status < 0) {
			$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
			return NULL;
		}
		if ($this->isBatched())
			return NULL;
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
	}
	
	public function remove($key) {
		$status = keyspace_client::Keyspace_Remove($this->co, $key);
		if ($status < 0) {
			$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
			return NULL;
		}
		if ($this->isBatched())
			return NULL;
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
		return $this->result->value();
	}
	
	public function rename($src, $dst) {
		$status = keyspace_client::Keyspace_Rename($this->co, $src, $dst);
		if ($status < 0) {
			$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
			return NULL;
		}
		if ($this->isBatched())
			return NULL;
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
	}

	public function prune($prefix) {
		$status = keyspace_client::Keyspace_Prune($this->co, $prefix);
		if ($status < 0) {
			$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
			return NULL;
		}
		if ($this->isBatched())
			return NULL;
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
	}	

	public function setExpiry($key, $exptime) {
		$status = keyspace_client::Keyspace_SetExpiry($this->co, $key, $exptime);
		if ($status < 0) {
			$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
			return NULL;
		}
		if ($this->isBatched())
			return NULL;
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
	}	

	public function removeExpiry($key) {
		$status = keyspace_client::Keyspace_RemoveExpiry($this->co, $key);
		if ($status < 0) {
			$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
			return NULL;
		}
		if ($this->isBatched())
			return NULL;
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
	}	

	public function begin() {
		return keyspace_client::Keyspace_Begin($this->co);
	}
	
	public function submit() {
		$status = keyspace_client::Keyspace_Submit($this->co);
		$this->result = new Result(keyspace_client::Keyspace_GetResult($this->co));
		return $status;
	}
	
	public function cancel() {
		return keyspace_client::Keyspace_Cancel($this->co);
	}
	
	public function isBatched() {
		return keyspace_client::Keyspace_IsBatched($this->co);
	}
	
	private function listArg($params, $name) {
		if (isset($params[$name]))
			return $params[$name];
			
		switch ($name) {
		case "prefix": return "";
		case "start_key": return "";
		case "count": return 0;
		case "skip": return FALSE;
		case "forward": return TRUE;
		}
	}
}

?>
